// @spotify/web-api-ts-sdk/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.emptyAccessToken = exports.Scopes = exports.DefaultResponseDeserializer = exports.DefaultResponseValidator = exports.DocumentLocationRedirectionStrategy = exports.NoOpErrorHandler = exports.ConsoleLoggingErrorHandler = exports.GenericCache = exports.LocalStorageCachingStrategy = exports.InMemoryCachingStrategy = exports.ClientCredentialsStrategy = exports.AuthorizationCodeWithPKCEStrategy = exports.SpotifyApi = void 0;
const SpotifyApi_js_1 = require("./SpotifyApi.js");
Object.defineProperty(exports, "SpotifyApi", { enumerable: true, get: function () { return SpotifyApi_js_1.SpotifyApi; } });
const AuthorizationCodeWithPKCEStrategy_js_1 = __importDefault(require("./auth/AuthorizationCodeWithPKCEStrategy.js"));
exports.AuthorizationCodeWithPKCEStrategy = AuthorizationCodeWithPKCEStrategy_js_1.default;
const ClientCredentialsStrategy_js_1 = __importDefault(require("./auth/ClientCredentialsStrategy.js"));
exports.ClientCredentialsStrategy = ClientCredentialsStrategy_js_1.default;
const InMemoryCachingStrategy_js_1 = __importDefault(require("./caching/InMemoryCachingStrategy.js"));
exports.InMemoryCachingStrategy = InMemoryCachingStrategy_js_1.default;
const LocalStorageCachingStrategy_js_1 = __importDefault(require("./caching/LocalStorageCachingStrategy.js"));
exports.LocalStorageCachingStrategy = LocalStorageCachingStrategy_js_1.default;
const GenericCache_js_1 = __importDefault(require("./caching/GenericCache.js"));
exports.GenericCache = GenericCache_js_1.default;
const ConsoleLoggingErrorHandler_js_1 = __importDefault(require("./errorhandling/ConsoleLoggingErrorHandler.js"));
exports.ConsoleLoggingErrorHandler = ConsoleLoggingErrorHandler_js_1.default;
const NoOpErrorHandler_js_1 = __importDefault(require("./errorhandling/NoOpErrorHandler.js"));
exports.NoOpErrorHandler = NoOpErrorHandler_js_1.default;
const DocumentLocationRedirectionStrategy_js_1 = __importDefault(require("./redirection/DocumentLocationRedirectionStrategy.js"));
exports.DocumentLocationRedirectionStrategy = DocumentLocationRedirectionStrategy_js_1.default;
const DefaultResponseValidator_js_1 = __importDefault(require("./responsevalidation/DefaultResponseValidator.js"));
exports.DefaultResponseValidator = DefaultResponseValidator_js_1.default;
const DefaultResponseDeserializer_js_1 = __importDefault(require("./serialization/DefaultResponseDeserializer.js"));
exports.DefaultResponseDeserializer = DefaultResponseDeserializer_js_1.default;
const Scopes_js_1 = require("./Scopes.js");
Object.defineProperty(exports, "Scopes", { enumerable: true, get: function () { return Scopes_js_1.Scopes; } });
const IAuthStrategy_js_1 = require("./auth/IAuthStrategy.js");
Object.defineProperty(exports, "emptyAccessToken", { enumerable: true, get: function () { return IAuthStrategy_js_1.emptyAccessToken; } });
